pkg/pma/wal.c

```diff
@@ -0,0 +1,498 @@
+/// @file
+///
+/// This file implements the WAL interface. It checks for corruption using a per
+/// page entry checksum, which is a checksum of the entry's page index and page
+/// contents, and a global checksum, which is the checksum composed of all per
+/// page entry checksums.
+
+#include "wal.h"
+
+#include <assert.h>
+#include <errno.h>
+#include <fcntl.h>
+#include <stdio.h>
+#include <stdlib.h>
+#include <string.h>
+#include <sys/stat.h>
+#include <unistd.h>
+
+#include "murmur3.h"
+#include "util.h"
+
+#ifdef FCNTL_FSYNC
+#    define fsync(fd) fcntl(fd, F_FULLFSYNC)
+#endif
+
+//==============================================================================
+// CONSTANTS
+
+/// Seed value for hashing function used to compute page checksum.
+static const size_t kSeed = 0;
+
+/// File stem for a WAL's data file.
+static const char kDataStem[] = "data";
+
+/// File stem for a WAL's metadata file.
+static const char kMetaStem[] = "meta";
+
+/// File extension for a WAL file.
+static const char kWalExt[] = "wal";
+
+/// Size in bytes of _metadata_entry_t's pg_idx field. Must be a macro to
+/// ensure kPageIdxSz + kPageSz can be resolved at compile time.
+#define kPageIdxSz sizeof(((_metadata_entry_t *)NULL)->pg_idx)
+
+//==============================================================================
+// TYPES
+
+/// An entry in a WAL's metadata file.
+typedef struct _metadata_entry {
+    /// Page index.
+    int64_t pg_idx;
+    /// Checksum of page index and page contents.
+    uint64_t checksum;
+} _metadata_entry_t;
```

Comment:
The motivation behind the single file design was simplicity, but that's since been changed. The metadata entries are--as of 1b0e5478954--actually no longer in the data file but instead in the metadata file, which, as you point out, is where they belong. The layout of the metadata file is:
```
global checksum
entry 0
entry 1
...
entry N
```
The layout of the data file is:
```
page 0
page 1
...
page N
```

I chose 64-bit integers over 32-bit integers for the sake of future proofing, say in case we want to use an 8-byte hash rather than a 4-byte hash. What added benefit do you see in unsigned page indexes + lengths instead of signed page indexes?

The WAL tests do exercise the corruption detection logic. There are tests for detecting a corrupted page in the data file, a corrupted page index in the metadata file, a corrupted entry checksum in the metadata file, and a corrupted global checksum in the metadata file. The first were introduced in 1b0e5478954, and the second two were introduced in a2d30db350 and 4902621e7b7a9cb6e9f59d8be4cc26ca732fd065.

I'll add a version number here shortly.